spec/consensus/light-client.md

```diff
@@ -1,318 +1,577 @@
 
-1. **Uninterrupted sequence of proof.** If a block is appended to the chain, where the last block
-   is trusted (and properly committed by the old validator set in the next block),
-   and the new block contains a new validator set, the new block is trusted if the lite client knows all headers in the prefix.
-   Intuitively, a trusted validator set is assumed to only chose a new validator set that will obey the Tendermint Failure Model.
+```go
+func verifySingle(trustedState TrustedState,
+                  untrustedSh SignedHeader,
+                  untrustedVs ValidatorSet,
+                  untrustedNextVs ValidatorSet,
+                  trustThreshold TrustThreshold) error {
+
+    untrustedHeader = untrustedSh.Header
+    untrustedCommit = untrustedSh.Commit
+
+    trustedHeader = trustedState.SignedHeader.Header
+    trustedVs = trustedState.ValidatorSet
+
+    assert trustedHeader.Height < untrustedHeader.Height AND
+           trustedHeader.Time < untrustedHeader.Time
+
+    // validate the untrusted header against its commit, vals, and next_vals
+    error = validateSignedHeaderAndVals(untrustedSh, untrustedVs, untrustedNextVs)
+    if error != nil return error
+
+    // check for adjacent headers
+    if untrustedHeader.Height == trustedHeader.Height + 1 {
+        if trustedHeader.NextValidatorsHash != untrustedHeader.ValidatorsHash {
+            return ErrInvalidAdjacentHeaders
+        }
+    } else {
+        error = verifyCommitTrusting(trustedVs, untrustedCommit, trustThreshold)
+        if error != nil return error
+    }
 
-2. **Trusting period.** Based on a trusted block _h_, and the lite client
-   invariant, which ensures the fault assumption during the trusting period, we can check whether at least one validator, that has been continuously correct from _h.Header.bfttime_ until now, has signed _snh_.
-   If this is the case, similarly to above, the chosen validator set in _snh_ does not violate the Tendermint Failure Model.
+    // verify the untrusted commit
+    return verifyCommitFull(untrustedVs, untrustedCommit)
+}
 
-3. **Bisection.** If a check according to the trusting period fails, the lite client can try to obtain a header _hp_ whose height lies between _h_ and _snh_ in order to check whether _h_ can be used to get trust for _hp_, and _hp_ can be used to get trust for _snh_. If this is the case we can trust _snh_; if not, we may continue recursively.
+// returns nil if header and validator sets are consistent; otherwise returns error
+func validateSignedHeaderAndVals(signedHeader SignedHeader, vs ValidatorSet, nextVs ValidatorSet) error {
+    header = signedHeader.Header
+    if hash(nextVs) != header.NextValidatorsHash OR
+        hash(vs) != header.ValidatorsHash OR
+        !matchingCommit(header, signedHeader.Commit) { return error }
+    return nil
+}
 
-## How to use it
+// returns nil if at least single correst signer signed the commit; otherwise returns error
```

Comment:
modulo the trustLevel ...

Also we're using this TrustThreshold type but it's not defined anywhere. Maybe we can keep it simple here and assert its a float between 1/3 and 2/3

Reply:
I added a short sentence making this assumption explicit.